sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -339,9 +339,16 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
     }
   }
 
-  private def assertNotBucketed(operation: String): Unit = {
-    if (numBuckets.isDefined || sortColumnNames.isDefined) {
-      throw new AnalysisException(s"'$operation' does not support bucketing right now")
```

Comment:
How about keeping the function name unchanged and just changing this message and list the sort columns if having. Something like:

> '$operation' does not support bucketing. Number of buckets: ...; sortBy: ...;

Reply:
I agree with you. 

I also found in `getBucketSpec`, when `numBuckets.isEmpty && sortColumnNames.isDefined`, it will throw `IllegalArgumentException`.

How about alternatively, we throw `AnalysisException` for all the cases for consistency?

```scala
  private def getBucketSpec: Option[BucketSpec] = {
    assertNotSortByOrBucketedBy()

    numBuckets.map { n =>
      BucketSpec(n, bucketColumnNames.get, sortColumnNames.getOrElse(Nil))
    }
  }

  private def assertNotSortByOrBucketedBy(): Unit = {
    if (sortColumnNames.isDefined && numBuckets.isEmpty) {
      throw new AnalysisException("sortBy must be used together with bucketBy")
    }
  }

  private def assertNotBucketedAndNotSorted(operation: String): Unit = {
    assertNotSortByOrBucketedBy()

    if (numBuckets.isDefined) {
      if (sortColumnNames.isDefined) {
        throw new AnalysisException(
          s"'$operation' does not support bucketBy and sortBy within a bucket right now")
      } else {
        throw new AnalysisException(s"'$operation' does not support bucketBy right now")
      }
    }
  }
```

Reply:
how about minimizing the code changes
```
private def assertNotBucketed(operation: String): Unit = {
  if (getBucketSpec.isDefined) {
    throw new AnalysisException(s"'$operation' does not support bucketing right now")
  }
}
```